### PR TITLE
Make `BTreeMap::clone()` not allocate when cloning an empty tree.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @Gankro 

CC @porglezomp

### CODE REVIEW

File: src/liballoc/btree/map.rs [hunk 1]
```diff
@@ -213,7 +213,16 @@ impl<K: Clone, V: Clone> Clone for BTreeMap<K, V> {
             }
         }
 
-        clone_subtree(self.root.as_ref())
+        if self.len() == 0 {
+            // XXX: ideally we'd call `BTreeMap::new` here, but that has the
```
Comment (Gankro) [inline src/liballoc/btree/map.rs]: Tidy is complaining XXX is deprecated notation; i would just drop the XXX, comment is fine without it

**Comment by Mark-Simulacrum** [comment]
@bors delegate+

(this is effectively r=me at this point pending rebase and nit)

**Pull request merged**